src/firecracker/examples/uffd/uffd_utils.rs

```diff
@@ -162,6 +162,110 @@ impl UffdPfHandler {
     }
 }
 
+#[derive(Debug)]
+pub struct Runtime {
+    stream: UnixStream,
+    backing_file: File,
+    backing_memory: *mut u8,
+    backing_memory_size: usize,
+    uffds: HashMap<i32, UffdPfHandler>,
+}
+
+impl Runtime {
+    pub fn new(stream: UnixStream, backing_file: File) -> Self {
+        let file_meta = backing_file
+            .metadata()
+            .expect("can not get backing file metadata");
+        let backing_memory_size = file_meta.len() as usize;
+        // # Safety:
+        // File size and fd are valid
+        let ret = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                backing_memory_size,
+                libc::PROT_READ,
+                libc::MAP_PRIVATE,
+                backing_file.as_raw_fd(),
+                0,
+            )
+        };
+        if ret == libc::MAP_FAILED {
+            panic!("mmap on backing file failed");
+        }
```

Comment:
Nit: A definite nit, feel free to ignore.

We can add some additional safety here by using https://docs.rs/nix/latest/nix/sys/mman/fn.mmap.html (nix is already in our dependency tree).

```suggestion
        let ret = unsafe { mmap(
            None,
            NonZeroUsize::new(backing_memory_size)?,
            ProtFlags::PROT_READ,
            MapFlags::MAP_PRIVATE,
            Some(backing_file),
            0
        )? };
```